docs/debugging.md

```diff
@@ -0,0 +1,179 @@
+# Uncovering concurrency issues, testing and debugging
+Until now, the GIL has allowed developers to elide and get away from
+concurrency issues when writing parallel programs, since the GIL ensured that
+all thread execution got serialized when any of them tried to simultaneously
+write or read a variable defined in the interpreter. Under the new free-threaded
+paradigm, developers now must think about concurrency, distribution and
+parallelism constructs that will allow them to exploit the maximum performance
+of their parallel programs.
+
+Usually, concurrency issues arise when two or more threads try to modify the
+same value in memory. In Python, this commonly occurs when a class or function
+defines shared state, either via an attribute or a variable that can modified
+from each thread execution scope.
+
+The most common issues related to concurrency in the context of free-threaded
+Python are either dirty reads/writes to data, unexpected behavior due to
+simultaneous access to native libraries that are not thread-safe, and finally,
+major runtime crashes due to memory allocation issues and forbidden
+pointer lookups. While the first case depends on the actual implementation of
+the algorithm/routine and may produce unintended results, it would not cause
+a fatal crash of the interpreter, as opposed to the last two cases.
+
+In order to discover, handle and debug concurrency issues at large, there are
+several strategies, which we will summarize next.
+
+## Identify shared state objects.
```

Comment:
I'm not sure this section is needed, since it is covered a lot more extensively in `porting.md`.

Reply:
Agreed, link to the porting guide for techniques to statically identify shared state in C code.